src/Controller/ResetPasswordController.php

```diff
@@ -71,63 +72,66 @@ public function checkEmail(SessionInterface $session): Response
     {
-        //Put token in session and redirect to self
         if ($token) {
-            // We store token in session and remove it from the URL,
-            // to avoid any leak if someone get to know the URL (AJAX requests, Analytics...).
+            // We store the token in session and remove it from the URL, to avoid the URL being
+            // loaded in a browser and potentially leaking the token to 3rd party JavaScript.
             $this->storeTokenInSession($token);
 
             return $this->redirectToRoute('app_reset_password');
         }
 
-        //Get token out of session storage
         $token = $this->getTokenFromSession();
         if (null === $token) {
             throw $this->createNotFoundException();
         }
 
-        //Validate token using password helper
-        $user = $this->resetPasswordHelper->validateTokenAndFetchUser($token);
+        try {
+            $user = $this->resetPasswordHelper->validateTokenAndFetchUser($token);
+        } catch (ResetPasswordExceptionInterface $e) {
+            $this->addFlash('reset_password_error', \sprintf(
+                'There was a problem validating your reset request - %s',
+                $e->getReason()
+            ));
 
-        //Reset password after token verified
-        $form = $this->createForm(PasswordResettingFormType::class);
+            return $this->redirectToRoute('app_forgot_password_request');
+        }
+
+        // The token is valid; allow the user to change their password.
+        $form = $this->createForm(ChangePasswordFormType::class);
         $form->handleRequest($request);
 
         if ($form->isSubmitted() && $form->isValid()) {
-            // A ResetPasswordToken should be used only once, remove it.
+            // A password reset token should be used only once, remove it.
             $this->resetPasswordHelper->removeResetRequest($token);
 
             // Encode the plain password, and set it.
-            //@TODO Encode password expects a UserInterface, we are only guaranteeing an object
             $encodedPassword = $passwordEncoder->encodePassword(
                 $user,
                 $form->get('plainPassword')->getData()
             );
 
-            /**
-             * We are assuming the user object HAS a setPassword method..
-             * @TODO handle this if it doesnt/sets password using different method.
-             * @psalm-suppress UndefinedInterfaceMethod
-             */
             $user->setPassword($encodedPassword);
             $this->getDoctrine()->getManager()->flush();
 
-            // TODO: please check the login route | CHANGE ROUTE, APP_FORGOT_PASSWORD_REQUEST USED IN DEVELOPMENT.
-            return $this->redirectToRoute('app_forgot_password_request');
+            // The session is cleaned up after the password has been changed.
+            $this->cleanSessionAfterReset();
```

Comment:
We are clean the session late in the game. This could also be done right after we validate token and fetch the user from persistence. 

Cleaning late may provide a better UX if there are hiccups. Cleaning early ensures the session is cleaned up right away. I can't think of any security issues off the top of my head for cleaning up late.

Reply:
Let's leave it - I can't think of any security issues. The critical thing is that we call `$this->resetPasswordHelper->removeResetRequest($token);` immediately.